src/ReverseProxy/Delegation/HttpSysDelegator.cs

```
@@ -190,69 +202,230 @@ private void RemoveDeadQueueReferences()
             _syncRoot = new object();
+            _ruleResetEvent = new ManualResetEventSlim(initialState: true);
+            _currentState = new DelegationQueueState();
+            _delegatingCount = new AtomicCounter();
+        }
+
+        public bool TryDelegate(
+            ILogger logger,
+            IServerDelegationFeature serverDelegationFeature,
+            IHttpSysRequestDelegationFeature requestDelegationFeature,
+            DestinationState destination,
+            [NotNullWhen(false)] out Exception? delegationException)
+        {
+            var state = _currentState;
+            var delegated = false;
+            delegationException = state.InitializationException;
+
+            if (state.IsInitialized)
+            {
+                delegated = TryDelegateInternal(requestDelegationFeature, state.Rule, _ruleResetEvent, _delegatingCount, out delegationException);
+                if (!delegated && ShouldTryToReset(delegationException, destination))
+                {
+                    state = Reset(serverDelegationFeature, state);
+
+                    Log.QueueReset(logger, destination, state.InitializationException);
+
+                    if (state.IsInitialized)
+                    {
+                        delegated = TryDelegateInternal(requestDelegationFeature, state.Rule, _ruleResetEvent, _delegatingCount, out delegationException);
+                    }
+                }
+            }
+
+            return delegated;
+
+            static bool TryDelegateInternal(
+                IHttpSysRequestDelegationFeature requestDelegationFeature,
+                DelegationRule rule,
+                ManualResetEventSlim resetEvent,
+                AtomicCounter delegatingCount,
+                [NotNullWhen(false)] out Exception? delegateException)
+            {
+                try
+                {
+                    // This ensures we don't dispose/reset the while a request is trying to delegate
```

I could have used a ReaderWriterLockSlim here instead. Since the most common case is there are lots of readers and very seldom a writer (i.e. a reset), this approach seemed like less overhead because in the no-writer case, it's just an if check inside of ManualResetEventSlim. ReaderWriterLockSlim is much more complex and involves a SpinLock every time.

I'm open to other ideas/suggestions.

---

The delegation uses SafeHandles that should prevent disposal while in use. Worse case you have a race where it's disposed just before use and you get an ODE you can catch here. You shouldn't need the MRES or ref counting AtomicCounter here, let SafeHandle take care of it.

---

I was originally thinking of just catching the ODE but due to the linked line below, it's possible that we unset the delegation property then before we dispose the safehandle another thread tries to delegate. At that point I think http.sys will error out but not sure what error code it will return.

https://github.com/dotnet/aspnetcore/blob/de3019b7dab5b5995942db1fbec6aa466c4af34c/src/Servers/HttpSys/src/DelegationRule.cs#L50

Assuming I figure out the error code http.sys would return in this case, what would you want to happen if an ODE is throws or the error code is returned? Retry or just let it fail?

---

Per our chat, removing all retries here now that we back ported the fix to ASP.NET Core 6. https://github.com/dotnet/aspnetcore/pull/40967